src/module.c

```diff
@@ -109,54 +110,118 @@ English.
 Returns OK on success, NOADD if the document was not added due to an IF expression not evaluating to
 true or an error if something went wrong.
 */
-static int doAddDocument(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, int canBlock) {
-  int nosave = RMUtil_ArgExists("NOSAVE", argv, argc, 1);
-  int fieldsIdx = RMUtil_ArgExists("FIELDS", argv, argc, 1);
-  int replace = RMUtil_ArgExists("REPLACE", argv, argc, 1);
-  int partial = RMUtil_ArgExists("PARTIAL", argv, argc, 1);
 
-  if (canBlock) {
-    canBlock = CheckConcurrentSupport(ctx);
+static int parseDocumentOptions(AddDocumentOptions *opts, ArgsCursor *ac, QueryError *status) {
+  // Assume argc and argv are at proper indices
+  int nosave = 0, replace = 0, partial = 0, foundFields = 0;
+  opts->fieldsArray = NULL;
+  opts->numFieldElems = 0;
+  opts->options = 0;
+
+  ACArgSpec specs[] = {{.name = "NOSAVE", .type = AC_ARGTYPE_BOOLFLAG, .target = &nosave},
+                       {.name = "REPLACE", .type = AC_ARGTYPE_BOOLFLAG, .target = &replace},
+                       {.name = "PARTIAL", .type = AC_ARGTYPE_BOOLFLAG, .target = &partial},
+                       {.name = "PAYLOAD", .type = AC_ARGTYPE_RSTRING, .target = &opts->payload},
+                       {.name = "LANGUAGE", .type = AC_ARGTYPE_STRING, .target = &opts->language},
+                       {.name = "IF", .type = AC_ARGTYPE_STRING, .target = &opts->evalExpr},
+                       {.name = NULL}};
+
+  while (!AC_IsAtEnd(ac)) {
+    int rv = 0;
+    const ACArgSpec *curSpec = NULL;
+    if ((curSpec = AC_ParseArgSpec(ac, specs, &rv))) {
```

Comment:
don't you think the specs should be defined in the args.h file on some other file as global var and not here inside a function?

Reply:
This is also confusing, so the function might returns NULL but this is not an error and we should handle the argument in another way? But if the rv indicate an error then this is an error? and what happened if it returns NULL and err? I think we should come with a better api here, maybe handle all the possibilities inside AC_ParseArgSpec